internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,311 @@
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents an HTTP/2 Frame.
+type Frame interface {
+	Header() *FrameHeader
+	// Free frees the underlying buffer if present so it can be reused by the
+	// framer.
+	Free()
+}
+
+// DataFrame is the representation of a [DATA frame]. It implements the Frame
+// interface. DATA frames convey arbitrary, variable-length sequences of octets
+// associated with a stream.
+//
+// [DATA frame]: https://httpwg.org/specs/rfc7540.html#DATA
+type DataFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *DataFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// HeadersFrame is the representation of a [HEADERS Frame]. It implements
+// the Frame interface. The HEADERS frame is used to open a stream, and
+// additionally carries a header block fragment.
+//
+// [HEADERS Frame]: https://httpwg.org/specs/rfc7540.html#HEADERS
+type HeadersFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// RSTStreamFrame is the representation of a [RST_STREAM Frame]. It implements
+// the Frame interface. There is no underlying byte array in this frame, so
+// Free() is a no-op.
+//
+// The RST_STREAM frame allows for immediate termination of a stream
+//
+// [RST_STREAM Frame]: https://httpwg.org/specs/rfc7540.html#RST_STREAM
+type RSTStreamFrame struct {
+	hdr  *FrameHeader
+	Code ErrCode
+}
+
+func (f *RSTStreamFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *RSTStreamFrame) Free() {}
+
+// SettingsFrame is the representation of a [SETTINGS Frame]. It implements
+// the Frame interface. There is no underlying byte array in this frame, so
+// Free() is a no-op.
+//
+// The SETTINGS frame conveys configuration parameters that affect how
+// endpoints communicate, such as preferences and constraints on peer behavior.
+//
+// [SETTINGS Frame]: https://httpwg.org/specs/rfc7540.html#SETTINGS
+type SettingsFrame struct {
+	hdr      *FrameHeader
+	Settings []Setting
+}
+
+func (f *SettingsFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *SettingsFrame) Free() {}
+
+// PingFrame is the representation of a [PING Frame]. It implements the Frame
+// interface. The PING frame is a mechanism for measuring a minimal round-trip
+// time from the sender, as well as determining whether an idle connection is
+// still functional.
+//
+// [PING Frame]: https://httpwg.org/specs/rfc7540.html#PING
+type PingFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *PingFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *PingFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// GoAwayFrame is the representation of a [GOAWAY Frame]. It implements the
```

Comment:
Instead of saying that the particular struct implements the framer interface in each of them, it would be nicer to state something like `// Each concrete frame type struct defined below implements the this interface` in the docstring of the Framer interface.